README.md
# Concrète Mixer

## About

Concrète Mixer is an ambient jukebox program. What this means is if you supply the software with a set of sound files, it will mix and scramble the sounds in interesting ways.

Concrète Mixer is designed to be run on a [Raspberry Pi](https://www.raspberrypi.org/). When a Pi is hooked up to loudspeakers your sounds can haunt any space of your choosing.

### What does it sound like?

Have a listen to an [hour long demo](https://concrete-mixer.bandcamp.com). This demo consists of two renderings of the app with the same set of sounds. The demo tracks are presented 'as-is' with no post-processing or editing.

### What's it written in?

The audio processing is written in [ChucK](http://chuck.cs.princeton.edu). A small shell script is used to initiate playback.

## Prerequisites

Concrète Mixer is intended for use on a Raspberry Pi 2 Model B running [Rasbpian](https://www.raspbian.org/). (Earlier Pi models *may* work with overclocking and special configuration settings applied.) In principle though, Concrète Mixer will run on any OSes/devices on which Perl and ChucK are available.

Finally, basic familiarity with running shell commands in linux is required.

## Installation and operation

1. The first thing you'll need is a set of sound files you want the software to mix. If you don't yet have a collection of sound files, you can [use the Concrète Mixer demo files](https://s3-us-west-1.amazonaws.com/concrete-mixer/concrete-mixer-files.zip). See [Tips](#tips) regarding using your own sound files.

2. Visit the [Concrète Mixer GitHub page](https://github.com/concrete-mixer/concrete-mixer) and click the [Download ZIP](https://github.com/concrete-mixer/concrete-mixer/archive/master.zip) link.

3. Unzip the code:
``$ unzip concrete-mixer-master.zip``

4. After unzipping you should have a directory called 'concrete-mixer-master'. Enter that directory and run the following commands to set up the config files:
``$ cp conf/global.conf.sample conf/global.conf``
``$ cp conf/concrete.conf.sample conf/concrete.conf``

5. Edit conf/concrete.conf and specify a directory location for your sounds:
``audioMainPath=<insert your dir here>``
Note that you can also supply a path for a second directory (audioAltPath) for sounds that you don't want to be played against each other; instead these sounds will be mixed with the 'main' sounds.

6. Run the app typing the following from your app's directory:
``./init.sh``

### Making a Raspberry Pi into a Concrète Mixer

The intention of Concrète Mixer is to turn a Pi into a single-purpose sound machine that may be left to run without any supervision indefinitely. You don't have to do this, but if you'd like to, here's what you do:

1. Edit conf/concrete.conf and set ``endlessPlay=1``. This will restart the app when it runs out of files to play.
2. To run Concrète Mixer each time the Pi is started, edit ``/etc/inittab`` using your favorite editor (here assuming nano):
    ``sudo nano /etc/inittab``
    Enter your password if required.
3. Search for the line ``1:2345:respawn:/sbin/getty 115200 tty1`` and comment it out by adding a ``#`` character at the start of line.
4. Add the following code beneath the commented out line: ``1:2345:respawn:/bin/login -f pi tty1 </dev/tty1 >/dev/tty1 2>&1``

    This line sets tty1 (the system's terminal number 1) to log in the pi user automatically on boot.

5. To get the Pi to run Concrète Mixer automatically, edit /home/pi/.bashrc:
    ``nano /home/pi/.bashrc``.

6. At the bottom of the file, add the following lines:
<code>
    if [ $(tty) == /dev/tty1 ]; then
        cd ~/concrete-mixer
        ./init.sh
    fi
</code>
This code will invoke Concrète Mixer if the current terminal is tty1. This means you can run the program in one terminal and use other terminals as required.

7. Save the file and restart the Pi. All going well, the Pi should start playing sound automatically after reboot.

## The art of Concrèting

### <a name="tips">Tips
* From experience sound files of about 90 seconds to two and a half minutes seem to work best in terms of the flow of the mix, but this will depend on the dynamics of the recording and (to a large degree) the taste of the listener.
* You should mix the samples' levels to be generally consistent so that any one sample should not be disproportionately louder than any other.
* You can specify several configuration options in the conf/concrete.conf file. Read concrete.conf.sample for more options.
* The Pi's analogue audio output is noisy; if possible use an HDMI audio splitter (preferably powered), or a USB sound card.
* The chuck executable file distributed with Concrète Mixer was compiled on a Raspberry Pi 2 Model B. This binary *may not* work on earlier or later Pi Models depending on their CPU architecture. If your Pi is incompatible, you can always download and compile a ChucK binary yourself; see the [ChucK download page](http://chuck.cs.princeton.edu/release/).

#### Running Concrète Mixer on other devices

You should be able to run Concrète Mixer GNU/Linux and OSX systems without much trouble as long as you have ChucK compiled and a bash shell; on Windows things should work as long as you can pass the ./concrete.ck file to ChucK in some way and the config file loads and file paths can be negotiated.
* [Information on how to install ChucK on various platforms](http://chuck.cs.princeton.edu/release)

Note that on other platforms the ChucK executable and chugin files included with Concrète Mixer will not work as it has been compiled for the Pi. To use Concrète Mixer on other platforms you'll need to change the concrete.conf file to point to a different ChucK executable (read the config file for more details).

## Licence

This code is distributed under the GPL v2. See the COPYING file for more details. The ChucK binary is also GPL v2.

## Acknowledgments

Many thanks to the [ChucK authors](http://chuck.cs.princeton.edu/doc/authors.html), especially for giving me their blessing to include the chuck binary with this program.

## Contact
<concretemixer.audio@gmail.com>


CHANGELOG.md
# Changelog

## 2.0
* Rewrite to remove the Perl wrapping, some tweaking of the way Fx libs work. Sound should be roughly the same.

## 1.0
* Initial release


init.pl
#!/usr/bin/perl

#----------------------------------------------------------------------------
#    Concrète Mixer - an ambient sound jukebox for the Raspberry Pi
#
#    Copyright (c) 2014 Stuart McDonald  All rights reserved.
#        https://github.com/concrete-mixer/concrete-mixer
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
#    U.S.A.
#----------------------------------------------------------------------------


use strict;
use warnings;

$ENV{PERL5LIB} = "./perllib";

use ConcreteMixer::Mode;
use Data::Dump qw(dump);

my $mode = ConcreteMixer::Mode->new;


init.sh
#    Concrète Mixer - an ambient sound jukebox for the Raspberry Pi
#
#    Copyright (c) 2014-2016 Stuart McDonald  All rights reserved.
#        https://github.com/concrete-mixer/concrete-mixer
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
#    U.S.A.

#!/bin/bash

ENDLESS_PLAY=0

# Before kicking off Concrete Mixer we need to know if we're up for endless
# play mode. To do this extremely cheaply and cheerfully, awk concrete.conf
if [ ! -f ./concrete.conf ]
then
    echo "concrete.conf is not available. Please read README.md for instructions"
else
    ENDLESS_PLAY=`awk 'match($0, /^endlessPlay=(1|0)$/) { print substr($0, 13, 1) }' ./concrete.conf`
fi

# By default, use version of chuck bundled with Concrete Mixer
# (compiled for rpi)
CHUCK_PATH=./chuck

# But for preference, use any locally installed version of chuck
if [ -x /usr/local/bin/chuck ]
then
    CHUCK_PATH=/usr/local/bin/chuck
fi

if [[ "$ENDLESS_PLAY" == 1 ]]
then
    # All going well, this will run eternally
    echo "Running Concrete Mixer for eternity, if possible"
    while [ true ]
    do
        $CHUCK_PATH concrete.ck || break
    done
else
    # Else, run just the once.
    echo "Running Concrete Mixer once"
    $CHUCK_PATH concrete.ck
fi


perllib/ConcreteMixer/Concrete.pm
#    Concrète Mixer - an ambient sound jukebox for the Raspberry Pi
#
#    Copyright (c) 2014 Stuart McDonald  All rights reserved.
#        https://github.com/concrete-mixer/concrete-mixer
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
#    U.S.A.

package ConcreteMixer::Concrete;
use base ConcreteMixer::Mode;
use strict;
use warnings;
use 5.10.0;
use List::Util;
use Data::Dump qw(dump);
my $data = {};

sub new {
    my ( $class, $config ) = @_;
    my $self = {};
    bless $self, $class;
    $self->{config} = $config;

    # do a bit of dying if config is nonsensical
    die "concurrent_sounds value not specified" if not $config->{concurrent_sounds};
    die "audio_main_path not specified" if not $config->{audio_main_path};

    if ( $config->{audio_alt_path} and $config->{concurrent_sounds} == 1 ) {
        $config->{concurrent_sounds} = 2;
        say "Setting concurrent_sounds to 2 as audio_alt_path specified";
    }

    $self->{play_files_main} = $self->get_files_list( $config->{audio_main_path} );

    die 'No main files present' if not scalar @{ $self->{play_files_main} };

    if ( $config->{audio_alt_path} ) {
        $self->{play_files_alt} = $self->get_files_list( $config->{audio_alt_path} );
        die 'No alt files present' if not scalar @{ $self->{play_files_alt} };
    }

    $self->{libpath} = 'lib/Modes/Concrete';
    $self->{fxChains} = $self->build_fxchains;
    $self->{playing_count} = 0;

    my $count = 0;

    while ( $count < $config->{concurrent_sounds} ) {
        $count++;

        if ( $self->{play_files_alt} and $count == $config->{concurrent_sounds} ) {
            $self->play_sound( 'alt' );
            next;
        }

        $self->play_sound;
    }

    if ( $config->{fx_chain_enabled} ) {
        my $fxChain = $self->get_fxchain;
        system( "$config->{chuck_path} + $self->{libpath}/playFxChain.ck:$fxChain" );
    }

    $data = $self;
    return $self;
}

# the OSC server callback
# processes messages back from playSound.ck and playFx.ck and respawns
# those processes until no sounds are left to play
sub process_osc_notifications {
    my ( $sender, $message ) = @_;
    my $self = $data;

    # first consider if we should serve the request or if there's
    # * a memory usage issue requiring process to end
    # * no more sounds to play and the last sound has finished playing
    if (
        $self->end_check # check for excess memory use
        # or if there's no sounds left to play
        or ( not $self->sounds_left and not $self->{playing_count} )
    ) {
        # don't do whatever you were going to do, fade out and either
        # end or reinitialise program
        $self->fade_mix;
    }

    # if we aren't already in a restart process or we've
    # discovered we need to kick off a restart process, carry on...
    if ( $message->[0] eq 'playSound' ) {
        my $type = $message->[2]; # should be 'main' or 'alt

        # decrement playing count as we know file has finished playing
        $self->{playing_count}--;

        if ( $self->sounds_left ) {
            $self->play_sound( $type );
        }
        else {
            $self->fade_mix;
        }
    }

    if ( $message->[0] eq 'playFxChain' ) {
        print "Got playFxChain notification, regenerating\n";
        my $fxChain = $self->get_fxchain;
        system( qq{$self->{config}{chuck_path} + $self->{libpath}/playFxChain.ck:$fxChain});
    }

    if ( $message->[0] eq 'fadeOutComplete' ) {
        $self->end;
    }
}

sub sounds_left {
    my ( $self ) = @_;

    # determine how many files are left
    my $count = scalar @{ $self->{play_files_main} };

    if ( defined $self->{play_files_alt} and scalar @{ $self->{play_files_alt} } ) {
        $count += scalar @{ $self->{play_files_alt} };
    };

    return $count;
}

sub end {
    my ( $self ) = @_;

    if ( $self->{config}{endless_play} ) {
        say "REINITIALISING\n";
        $self->reinitialise();
    }
    else {
        $self->kill_master_pid();
        say "EXITING";
        exit;
    }
}

=head2

playFx isn't playSound-aware, so when we run out of sounds we need to shut down the app manually.

This function invokes fadeMix, which zeroes the playback volume, and reports back via OSC, at which point we shut down or restart chuck.

=cut
sub fade_mix {
    my ( $self ) = @_;
    system( qq{$self->{config}{chuck_path} + $self->{libpath}/fadeMix.ck});
}

=head2 build_fxchains

Builds an random array of fx ids for use by playFx.ck. Once
the array is emptied of fx patches it gets rebuilt.

The point of doing it like this that we don't reuse an fxchain
until we've used all the others, meaning less repetition overall.

=cut
sub build_fxchains {
    my $self = shift;
    my @arr;

    # there are some fx we don't want rpis to use as they use chugens.
    # It's unfortunate we have to hardcode things here but then the
    # whole business of using Perl to bootstrap Chuck is unfortunate.
    if ( $self->{config}{rpi} ) {
        @arr = ( 1..19 );
    }
    else {
        @arr = ( 1..25 );
    }

    @arr = List::Util::shuffle @arr;

    return \@arr;
}

sub get_fxchain {
    my $self = shift;

    if ( not scalar @{ $self->{fxChains} } ) {
        $self->{fxChains} = $self->build_fxchains;
    }

    return pop @{ $self->{fxChains} };
}

sub play_sound {
    my ( $self, $type) = @_;

    $type //= 'main';

    chdir $self->{config}{cwd};
    my $filename;

    # get file to play
    # in the normal run of things, play_sound only gets called
    # when we know there's at least one file to play
    if ( $type eq 'alt' ) {
        $filename = pop @{ $self->{play_files_alt} };

        # if we've run out of files, try the 'main' pool
        if ( not $filename ) {
            $self->play_sound;
            return;
        }
    }
    else {
        $filename = pop @{ $self->{play_files_main} };

        # if we've run out of files, try the 'alt' pool
        if ( not $filename ) {
            $self->play_sound('alt');
            return;
        }
    }


    print "playSound playing $filename\n";
    my $command = "$self->{config}{chuck_path} + $self->{libpath}/playSound.ck:" . '"' . $filename . '"';

    if ( $type eq 'alt' ) {
        $command .= ":alt";
    }

    system( $command );
    $self->{playing_count}++;
}

1;


perllib/ConcreteMixer/Mode.pm
#    Concrète Mixer - an ambient sound jukebox for the Raspberry Pi
#
#    Copyright (c) 2014 Stuart McDonald  All rights reserved.
#        https://github.com/concrete-mixer/concrete-mixer
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
#    U.S.A.

package ConcreteMixer::Mode;

use 5.10.0;
use strict;
use warnings;
use List::Util;
use Data::Dump qw(dump);
use List::Util;
use Cwd;
use Net::OpenSoundControl::Server;
use Sys::Statistics::Linux::MemStats;
use Sys::Statistics::Linux::Processes;

sub new {
    my $class = shift;
    my $self = {};
    bless $self, $class;

    $self->initialise;

    return $self;
}

sub initialise {
    my ( $self, $reincarnation ) = @_;
    my $config = $self->get_config('conf/global.conf');

    # determine if we have minimum settings to kick things off
    if (
        not defined $config->{chuck_path}
        or not defined $config->{mode}
    ) {
        die "Insufficient settings to run program. Does config file exist and contain settings? See sample-config for details";
    }

    my $mconfig = $self->get_config( qq{conf/$config->{mode}.conf} );

    foreach my $key ( keys %{ $mconfig } ) {
        $config->{ $key } = $mconfig->{ $key };
    }

    $config->{cwd} = cwd();
    $self->{config} = $config;

    # first initialise chuck environment
    my $bufsize = $config->{bufsize} || 4096;
    my $srate = $config->{srate} || 44100;

    # kick off chuck loop vm
    my $pid = system( qq{$config->{chuck_path} --loop --srate$srate --bufsize$bufsize &} );

    # sleeps give time for chuck to initliaise
    sleep 1;
    my $mode = ucfirst( $config->{mode} );

    my $chuck_string = qq{ $config->{chuck_path} --chugin-path:$config->{chugin_path} + lib/Modes/$mode/$mode.ck:"$config->{bpm}":"$srate":"$config->{record}":"$config->{rpi}" };
    # say $chuck_string;
    system( $chuck_string );

    # as above
    sleep 1;
    my $mod_name = "ConcreteMixer::$mode";
    my $req_name = "perllib/ConcreteMixer/$mode.pm";
    require "$req_name"; 1;
    my $obj = $mod_name->new( $config );
    $self->{mode} = $obj;

    # OSC server can't be turned off, so don't start it again
    if ( not $reincarnation ) {
        $self->start_osc_server;
    }
}

sub start_osc_server {
    my $self = shift;
 
    $self->{osc_server} = Net::OpenSoundControl::Server->new(
        Port => 3141,
        Handler => $self->{mode}->can('process_osc_notifications'),
    ) or die "Could not start OSC server: $@\n";

    $self->{osc_server}->readloop();
}

sub reinitialise {
    my $self = shift;

    $self->kill_master_pid();

    # arg denotes 'reincarnation' rather than true initialisation
    $self->initialise( 1 );
}

sub kill_master_pid {
    my $self = shift;

    my $chuck_master_pid = $self->get_chuck_master_pid();

    my $result = `kill $chuck_master_pid`;

    # killing should happen silently
    if ( $result ) {
        die "Could not stop chuck process: $!";
    }
}

sub get_files_list {
    my ( $self, $path ) = @_;
    my $glob_target = $path . '/*.wav';
    my @source_files = glob( $glob_target );
    @source_files = List::Util::shuffle( @source_files );
    return \@source_files;
}

sub get_config {
    my ( $self, $path ) = @_;
    my $config = {};

    open( my $fh, $path ) or die "Cannot open config file $path";
    my @config_rows = <$fh>;

    foreach my $row ( @config_rows ) {
        next if ( $row =~ /^(#|\n$)/ );
        chomp $row;
        my ( $key, $value ) = split( '=', $row );
        $config->{ $key } = $value;
    }

    close $fh;
    return $config;
}

sub end_check {
    my $self = shift;

    my $lxsm  = Sys::Statistics::Linux::MemStats->new;
    my $mem_stats = $lxsm->get;

    # need to get values to common factor for comparison
    # easiest seems to be choose k(ibi)bytes, even if mem_stats
    # are originally calculated as mega rather than mibi bytes
    my $realfree = $mem_stats->{realfree} * 1024;
    my $memtotal = $mem_stats->{memtotal} * 1024;

    my $page_size = `getconf PAGE_SIZE`;

    my $chuck_master_pid = $self->get_chuck_master_pid();

    my $lxsp = Sys::Statistics::Linux::Processes->new(
        pages_to_bytes => $page_size,
        pids => [ $chuck_master_pid ]
    );

    $lxsp->init;
    my $stat = $lxsp->get;
    my $chuck_resident_mem_used = $stat->{ $chuck_master_pid }{resident};

    # The app leaks memory. The code has been rationalised somewhat,
    # but leakage seems to be increasing at a rate of about 1MB per playSound
    # and playFxChain cycle. I'm not sure why this is happening, but I'm guessing
    # the persistent use of the sound bus in the Control class is the source.
    # Independent instances of playSound do not leak, for example.
    #
    # The ChucK mailing list and documentation is vague about memory management
    # and garbage collection, but it looks like this issue isn't going to be
    # resolved any time soon, so we'll have to go with outside intervention to
    # manage memory leakage
    #
    # One of the primary goals of Concrète Mixer is for it to run indefinitely, so
    # we need to ensure that the memory leakage doesn't bring down the system.
    #
    # There are three scenarios in which Concrète Mixer should be restarted:
    # 1. The sample queue has been exhausted (as a polite housekeeping
    #   exercise, not managed here)
    # 2. The chuck process itself has used more than 50% of system memory
    # 3. More than 80% of memory has been utilised (all processed), to stop
    # chuck bringing down the system.
    #
    # Concrète Mixer is intended as a single use device (ie, nothing other than
    # system processes should be utilising resources) but it seems polite
    # nontheless to assume a limit of 50% of total resources, in the case
    # where a user decides to utilise some other software on their system - it is
    # their system. 50% memory equates to 256M on a Model B Pi, and 128M on a Model A. Unfortunately restarts are more
    # likely with longer samples, as samples lengths plus leakage are more likely
    # to tip the threshold.

    # Scenario no.2
    if ( $memtotal / 2 < $chuck_resident_mem_used ) {
        print "ending Concrète Mixer (free memory below 20%)\n";
        return 1;
    }

    # Scenario no. 3
    if ( $realfree * 5 < $memtotal ) {
        print "ending Concrète Mixer (chuck process using more than 50% system memory)\n";
        return 1;
    }

    return 0;
}

sub get_chuck_master_pid {
    my @process_results = `ps aux | grep 'chuck --loop'`;

    foreach my $process ( @process_results ) {
        my @frags = split(/\s+/, $process);

        if ( $frags[ -3 ] eq '--loop' ) {
            return $frags[1];
        }
    }
}

1;
